Userland/Libraries/LibGL/GLMat.cpp

```diff
@@ -0,0 +1,141 @@
+        break;
+    case GL_MODELVIEW:
+        g_gl_state->model_view_matrix_stack.append(g_gl_state->model_view_matrix);
+        break;
+    default:
+        dbgln("glPushMatrix(): Attempt to push matrix with invalid matrix mode {})", g_gl_state->curr_matrix_mode);
+        break;
+    }
+}
+
+/*
+ * Pop a matrix from the corresponding matrix stack into the
+ * corresponding matrix in the state based on the current
+ * matrix mode
+ */
+void glPopMatrix()
+{
+#ifdef GL_DEBUG
+    dbgln("glPopMatrix(): Popping matrix from matrix stack (matrix_mode = {})", g_gl_state->curr_matrix_mode);
+#endif
+
+    // TODO: Make sure stack::top() doesn't cause any  nasty issues if it's empty (that could result in a lockup/hang)
+    switch (g_gl_state->curr_matrix_mode) {
+    case GL_PROJECTION:
+        g_gl_state->projection_matrix = g_gl_state->projection_matrix_stack.take_last();
+        ;
+
+        break;
+    case GL_MODELVIEW:
+        g_gl_state->model_view_matrix = g_gl_state->model_view_matrix_stack.take_last();
+        ;
+        break;
+    default:
+        dbgln("glPopMatrix(): Attempt to pop matrix with invalid matrix mode, {}", g_gl_state->curr_matrix_mode);
+        break;
+    }
+}
+
+void glLoadIdentity()
+{
+    if (g_gl_state->curr_matrix_mode == GL_PROJECTION)
+        g_gl_state->projection_matrix.load_identity();
+    else if (g_gl_state->curr_matrix_mode == GL_MODELVIEW)
+        g_gl_state->model_view_matrix.load_identity();
+    else
+        VERIFY_NOT_REACHED();
+}
+
+/**
+ * Create a viewing frustum (a.k.a a "Perspective Matrix") in the current matrix. This
+ * is usually done to the projection matrix. The current matrix is then multiplied
+ * by this viewing frustum matrix.
+ *
+ * https://www.khronos.org/registry/OpenGL-Refpages/gl2.1/xhtml/glFrustum.xml
+ *
+ *
+ * TODO: We need to check for some values that could result in a division by zero
+ * or some other fuckery that will fuck with the driver here!
+ */
+void glFrustum(GLdouble left, GLdouble right, GLdouble bottom, GLdouble top, GLdouble nearVal, GLdouble farVal)
+{
+    Mat4 frustum;
+    GLdouble a;
+    GLdouble b;
+    GLdouble c;
+    GLdouble d;
+
+    // Let's do some math!
+    a = (right + left) / (right - left);
+    b = (top + bottom) / (top - bottom);
+    c = -((farVal + nearVal) / (farVal - nearVal));
+    d = -((2 * (farVal * nearVal)) / (farVal - nearVal));
+
+    frustum(0, 0, ((2 * nearVal) / (right - left)));
+    frustum(1, 1, ((2 * nearVal) / (top - bottom)));
+    frustum(2, 0, a);
+    frustum(2, 1, b);
+    frustum(2, 2, c);
+    frustum(2, 3, -1);
+    frustum(3, 2, d);
+    frustum(3, 3, 0);
+    // Phew
+
+    if (g_gl_state->curr_matrix_mode == GL_PROJECTION) {
+        g_gl_state->projection_matrix = g_gl_state->projection_matrix * frustum;
+    } else if (g_gl_state->curr_matrix_mode == GL_MODELVIEW) {
+#ifdef GL_DEBUG
+        dbgln("glFrustum(): frustum created with curr_matrix_mode == GL_MODELVIEW!!!");
+#endif
```

Comment:
Nit: This should be `dbgln_if(GL_DEBUG, "glFrustum(): frustum created with curr_matrix_mode == GL_MODELVIEW!!!");`

Reply:
Advice: if your IDE supports regex replace use
```regex
^#if (.*)\n( *)dbgln\((.*)\);\n#endif$
```
to search and
```regex
$2dbgln_if($1, $3);
```
to replace

this only works when GL_DEBUG is standard defined as 0 as all other DEBUG flags, see @ccapitalK 's next review